### PR TITLE
Draw selection preview around window

### LINKED ISSUE #8: Draw a selection preview around windows
Since a single click will return the clicked window's geometry, it would be nice to draw a rectangle around it like slop.

### PR DESCRIPTION
**Pull request by pickfire**
Fix #8

### CODE REVIEW

**Comment by Lonami** [comment]
Hm… these changes seem to regress rendering by *a lot*. The guides lag behind the mouse so much, at least on my machine.

File: src/main.rs [hunk 1]
```diff
@@ -95,15 +107,18 @@ fn main() -> Result<(), String> {
 
     if !opt.no_guides {
```
Comment (tnjd) [inline src/main.rs]: This new code should probably not be inside this if block, since no_guides is a different thing to no window selection preview

File: src/main.rs [hunk 2]
```diff
@@ -150,11 +165,22 @@ fn main() -> Result<(), String> {
                 let width = (right_x - left_x) as u16;
                 let height = (bottom_y - top_y) as u16;
 
+                let cursor_pt = xcb::Point::new(motion.event_x(), motion.event_y());
+                let result = get_window_at_point(
```
Comment (tnjd) [inline src/main.rs]: Seems like this call in the loop is likely to be what's causing the slowdown. Maybe we can cache the result, since we can know the window's geometry, and the mouse's new coordinates.

File: src/main.rs [hunk 3]
```diff
@@ -184,13 +210,22 @@ fn main() -> Result<(), String> {
...
+                        cursor_pt,
+                        opt.remove_decorations,
+                        Some(window),
                     );
-
-                    set_shape(&conn, window, &rects);
+                    let guides = build_guides(screen_rect, cursor_pt, guide_width);
+                    if let Some(result) = result {
+                        let borders = build_border(result.rect, line_width);
+                        let mut rects = guides.to_vec();
```
Comment (tnjd) [inline src/main.rs]: We could probably construct the new array manually:             let rects = [guides[0], guides[1], borders[0], borders[1], borders[2], borders[3]];

**Review by tnjd — CHANGES_REQUESTED** [review]
Thanks for doing this! I think the width / colour of the new lines warrant their own config flags, and maybe one for enabling/disabling them too.

File: src/lib/mod.rs [hunk 4]
```diff
@@ -206,13 +206,15 @@ pub fn get_window_at_point(
...
     remove_decorations: u32,
+    ignore_window: Option<xcb::Window>,
 ) -> Option<HacksawResult> {
     let tree = xcb::query_tree(conn, win).get_reply().unwrap();
     let children = tree
         .children()
         .iter()
         .filter(|&child| viewable(conn, *child))
         .filter(|&child| input_output(conn, *child))
+        .filter(|&child| ignore_window.map_or(true, |window| *child != window))
```
Comment (tnjd) [inline src/lib/mod.rs]: This line is a little hard to follow - could we make it more clear?

File: src/main.rs (hunk 2 above)
Comment (Lonami) [reply src/main.rs]: The problem with caching is that the user can change the windows at any time (or even the windows themselves can change on their own). My original idea was to query all windows to avoid the slowdown was to query all windows at startup and do the bounds check from within rust, but then the mentioned problem showed up.

File: src/main.rs (hunk 1 above)
Comment (pickfire) [reply src/main.rs]: So what do you think the behavior should do?

File: src/main.rs (hunk 2 above)
Comment (pickfire) [reply src/main.rs]: I did thought about this but it may change so we need to catch other events.

Comment (9ary) [reply src/main.rs]: The correct way to handle this would be to fetch the tree once and then listen to events to update it cheaply.

File: src/main.rs (hunk 3 above)
Comment (pickfire) [reply src/main.rs]: Yes, I thought about this. Too bad no `const-generics` for `[arr1, arr2].concat()`.

File: src/lib/mod.rs (hunk 4 above)
Comment (pickfire) [reply src/lib/mod.rs]: Do you have any idea how to make it clearer?

File: src/main.rs (hunk 2 above)
Comment (pickfire) [reply src/main.rs]: Yes, that's true.

**Comment by Lonami** [comment]
This does the trick for me:
```diff
diff --git a/src/main.rs b/src/main.rs
index edff83a..1eb0143 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -131,6 +131,10 @@ fn main() -> Result<(), String> {
     let mut in_selection = false;
     let mut ignore_next_release = false;
 
+    use std::time::{Instant, Duration};
+    let mut last_draw = Instant::now();
+    const DRAW_DELTA: Duration = Duration::from_millis(8);
+
     // TODO draw rectangle around window under cursor
     loop {
         let ev = conn
@@ -158,6 +162,12 @@ fn main() -> Result<(), String> {
                 return Err("Exiting due to ESC key press".into());
             }
             xcb::MOTION_NOTIFY => {
+                let now = Instant::now();
+                if now - last_draw < DRAW_DELTA {
+                    continue;
+                }
+                last_draw = now;
+
                 let motion: &xcb::MotionNotifyEvent = unsafe { xcb::cast_event(&ev) };
 
                 let (left_x, right_x) = min_max(motion.event_x(), start_pt.x());
```

Not that's the best code but it works.

**Comment by pickfire** [comment]
@Lonami That should be in another patch from what I will say since it fixes another problem.

**Comment by tnjd** [comment]
by the way @Lonami in your patch, your subtraction of Instants should be saturating_sub since I think they panic if they go negative

File: src/lib/mod.rs (hunk 4 above)
Comment (Lonami) [reply src/lib/mod.rs]: Perhaps explicitly `if let Some(w) = ignore_window { *child != window } else { true }`. But `map_or` is kinda nicer if you know what it does.

**Comment by Lonami** [comment]
> they panic if they go negative

They can't be negative because [`Instant`](https://doc.rust-lang.org/stable/std/time/struct.Instant.html) is monotonic:

> A measurement of a monotonically nondecreasing clock.

**Comment by pickfire** [comment]
@Lonami But rust does not know that, bound checks might still be there.

File: src/main.rs (hunk 2 above)
Comment (Lonami) [reply src/main.rs]: @9ary because it changes the way we currently do some things, it may deserve another PR.

**Comment by tnjd** [comment]
> > they panic if they go negative
> 
> They can't be negative because [`Instant`](https://doc.rust-lang.org/stable/std/time/struct.Instant.html) is monotonic:
> 
> > A measurement of a monotonically nondecreasing clock.


Ah you're right, I was confusing it with a different situation

**Comment by Lonami** [comment]
> @Lonami That should be in another patch

Most likely because it affects previous code, I only mentioned it because it makes this change-set usable.

**Comment by Lonami** [comment]
> @Lonami But rust does not know that, bound checks might still be there.

That's micro-optimization compared to the rest of things we're doing :)

File: src/main.rs (hunk 1 above)
Comment (tnjd) [reply src/main.rs]: Either move the window border drawing outside (ie do it regardless of whether the user wants guides) or with a separate no_window_border flag

File: src/main.rs (hunk 2 above)
Comment (tnjd) [reply src/main.rs]: As we discussed outside, the tree approach might add lot of complexity. A simpler approach might be (as suggested by novie) to throttle how often we query the tree, and keep working while we wait for the response. X is asynchronous after all. The simplest approach is the one Lonami sent a patch for, but it feels a bit dirty.

Comment (Lonami) [reply src/main.rs]: @expectocode other only approach I can think of is kind of spawn a timer and reset it on every move. When the timer is over, *then* we render. But that sounds a lot more complex.

File: src/main.rs (hunk 1 above)
Comment (pickfire) [reply src/main.rs]: Okay, will move it outside.

File: src/lib/mod.rs (hunk 4 above)
Comment (pickfire) [reply src/lib/mod.rs]: That's exactly the same as `map_or`.

Comment (Lonami) [reply src/lib/mod.rs]: Yes, I know, but I put it because @expectocode wanted the same behaviour in a "less confusing way". We should wait on their thoughts.

Comment (tnjd) [reply src/lib/mod.rs]: We could do `filter(|&child| Some(*child) != ignore_window)`

Comment (tnjd) [reply src/lib/mod.rs]: to remove the weird `*child`, we could take `ignore_window: Option<&xcb::Window>`

Comment (Lonami) [reply src/lib/mod.rs]: "Keep unless `ignore_window` is `Some(*child)`", so if it's `None` it's kept as well. Eh, takes a bit of thought still but I do find the `map_or(true, ...)` a bit weird.

Comment (tnjd) [reply src/lib/mod.rs]: I think it's likely that we'll change the type of ignore_window Option<Window> to just Window in the near future tbh, so this will all be moot

File: src/main.rs (hunk 2 above)
Comment (pickfire) [reply src/main.rs]: @Lonami Yes, I think it deserves another PR.

File: src/lib/mod.rs (hunk 4 above)
Comment (pickfire) [reply src/lib/mod.rs]: Ah, I didn't even thought about that, yeah `Some(*child) != ignore_window` is definitely better.

**Comment by tnjd** [comment]
This TODO should be removed: https://github.com/neXromancers/hacksaw/pull/22/files#diff-639fbc4ef05b315af92b4d836c31b023R140  (`// TODO draw rectangle around window under cursor`)

**Review by Lonami — APPROVED** [review]
I'm OK with merging these and then improving performance.

**Comment by pickfire** [comment]
@Lonami That's what I think it's best, just don't release before we update the other parts, I would also like to do more refactorings, especially the state of current functions. Also, I will plan out a public API for libraries then if I have the time I can help out improving the performance.

Note that there is a bug here when keys are pressed, windows changes. The border still stick to the old changes, we should listen to keyboard events now too but propagate them, I am not sure how to do this yet.

**Comment by tnjd** [comment]
@pickfire How about we create a branch to work on this, and merge it into master when it's performant and the bugs are worked out?

**Comment by 9ary** [comment]
I like this idea personally.

**Comment by pickfire** [comment]
@expectocode Of course, that's good.

**Comment by Lonami** [comment]
The most debated PR on the history of neXromancers so far has been merged 🎉
